### ISSUE TITLE
azuredevops_users datasource return id that can't be used in other resources 

### ISSUE THREAD

**Issue by patr00n** [microsoft/terraform-provider-azuredevops#400]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Affected Resource(s)

Data Source
* `azuredevops_users`

### Terraform Configuration Files

```hcl
data "azuredevops_project" "project" {
  name = "#####"
}

data "azuredevops_users" "user" {
  principal_name = #########@###.com"
}

resource "azuredevops_branch_policy_auto_reviewers" "policy" {
  project_id = data.azuredevops_project.project.id
  enabled  = true
  settings {
    auto_reviewer_ids  = tolist([data.azuredevops_users.user.id])
    scope {
      repository_id  = 00000000-0000-0000-0000-000000000000"
      repository_ref = "refs/heads/master"
      match_type     = "Exact"
    }
  }
}
```

### Debug Output

```
╷
│ Error: Error creating policy in Azure DevOps: TF402457: The settings for this policy are not correctly formatted. Error: Error converting value "users#dr84IrOhg6kMbx6wbJva58HcnqY=" to type 'System.Guid'. Path 'requiredReviewerIds[0]', line 6, position 40.
│ 
│   with azuredevops_branch_policy_auto_reviewers.policy,
│   on moonlight.tf line 151, in resource "azuredevops_branch_policy_auto_reviewers" "policy":
│  151: resource "azuredevops_branch_policy_auto_reviewers" "policy" {
```
### Expected Behavior

The data source outputs proper GUID
### Steps to Reproduce

1. `terraform apply`

### References

* #0000

**Comment by xuzhang3** [comment]
@patr00n `azuredevops_users` will read all t he users match the `principal_name` the `azuredevops_users` ID is meanness here.  You can find the user in `data.azuredevops_users.user.users`.   The problem here is service will return the user descriptor or user origin  ID but no user ID, a workaround here is add the user to a group and use this group as the auto reviewer.
```hcl
data "azuredevops_users" "user" {
  principal_name = "test@test.com"
}

data azuredevops_git_repository "r" {
  project_id = data.azuredevops_project.project.id
  name = "myproject"
}

resource "azuredevops_group" "g" {
  scope        = data.azuredevops_project.project.id
  display_name = "Branch Reviewer"
  members = [
    tolist(data.azuredevops_users.user.users)[0].descriptor
  ]
}

output "dd" {
  value = azuredevops_group.g
}

resource "azuredevops_branch_policy_auto_reviewers" "policy" {
  project_id = data.azuredevops_project.project.id
  enabled = true
  settings {
    auto_reviewer_ids = [azuredevops_group.g.origin_id]
    scope {
      repository_id = data.azuredevops_git_repository.r.id
      repository_ref = "refs/heads/master"
      match_type = "Exact"
    }
  }
}
```

**Comment by patr00n** [comment]
@xuzhang3  `data.azuredevops_users.user.users` got it, thanks!

>  The problem here is service will return the user descriptor or user origin ID but no user ID

This is ADO API behavior? Can we fix this?
And thanks for the workaround but it's not convenient for us because our groups are managed on an organization level and I can't modify them just read because I have admin rights only on my specific project. I think this is not only my case and it would be great to get proper id from data source.

**Comment by xuzhang3** [comment]
@patr00n The API  [Get User](https://docs.microsoft.com/en-us/rest/api/azure/devops/graph/users/get?view=azure-devops-rest-5.1) used in this resource can get the user info with descriptor but no user ID.  In v6.x, there an API [identities](https://docs.microsoft.com/en-us/rest/api/azure/devops/ims/identities/read%20identities?view=azure-devops-rest-6.0) can be used to get the user ID.  I think there must be other APIs that can be used to fetch the user ID.

**Comment by patr00n** [comment]
> @patr00n The API [Get User](https://docs.microsoft.com/en-us/rest/api/azure/devops/graph/users/get?view=azure-devops-rest-5.1) used in this resource can get the user info with descriptor but no user ID. In v6.x, there an API [identities](https://docs.microsoft.com/en-us/rest/api/azure/devops/ims/identities/read%20identities?view=azure-devops-rest-6.0) can be used to get the user ID. I think there must be other APIs that can be used to fetch the user ID.

So can I hope that the new resource will be added in one of the new provider versions?

**Comment by fdebuire** [comment]
@xuzhang3 The API [Get User](https://docs.microsoft.com/en-us/rest/api/azure/devops/graph/users/get?view=azure-devops-rest-5.1) returns a link that can be used to fetch the userId:
```json
"_links": {
    ...
    "storageKey": {
      "href": "https://vssps.dev.azure.com/Fabrikam/_apis/Graph/StorageKeys/aad.MDA0NzBlMzQtZGE2MS03YTY5LWJkOTYtNDg3YTg0OWVjNTU4"
    }
    ...
  },
```

When I add auto reviewers through the UI, that's exactly what is done.

**Comment by fdebuire** [comment]
For now I used this workaround:

```hcl
data "azuredevops_users" "aad-user" {
  for_each       = toset(local.reviewers)
  principal_name = format("%s", each.key)
}

data "external" "real-user-id" {
  for_each = toset(local.reviewers)
  program  = ["bash", "fetch-user-id.sh"]

  query = {
    descriptor = tolist(data.azuredevops_users.aad-user[each.key].users)[0].descriptor
  }
}

resource "azuredevops_branch_policy_auto_reviewers" "branch-policy-auto-reviewers" {
  project_id = "XXX"

  enabled  = true
  blocking = false

  settings {
    auto_reviewer_ids  = [for data in data.external.real-user-id : data.result["user_id"]]
    submitter_can_vote = false

    scope {
      repository_ref = "refs/heads/master"
      match_type     = "Exact"
    }
  }
}
```

```sh
#!/usr/bin/env bash

# see https://registry.terraform.io/providers/hashicorp/external/latest/docs/data-sources/data_source

# Exit if any of the intermediate steps fail
set -e

# Extract "descriptor" argument from the input into DESCRIPTOR shell variable.
# jq will ensure that the values are properly quoted
# and escaped for consumption by the shell.
eval "$(jq -r '@sh "DESCRIPTOR=\(.descriptor)"')"

user_id=$(curl --silent \
  -u ":$AZDO_PERSONAL_ACCESS_TOKEN" \
  "https://vssps.dev.azure.com/xxxxx/_apis/Graph/StorageKeys/$DESCRIPTOR" | jq -r ".value")

# Safely produce a JSON object containing the result value.
# jq will ensure that the value is properly quoted
# and escaped to produce a valid JSON string.
jq -n --arg user_id "$user_id" '{"user_id":$user_id}'
```

**Comment by xuzhang3** [comment]
@fdebuire  fixed in #482

**Comment by drdamour** [comment]
@xuzhang3 this has made things TERRIBLY slow in my environment.  it takes a good 5 minutes to get all this extra data for users.  also ran into #492
